File: dotnet/xmsclient/Client/State/AMQStateManager.cs
using System;
using System.Collections;
using jpmorgan.mina.common;
using log4net;
using OpenAMQ.Framing;
using OpenAMQ.XMS.Client.Protocol;

namespace OpenAMQ.XMS.Client.State
{
    public class AMQStateManager : IAMQMethodListener
    {
        private static readonly ILog _logger = LogManager.GetLogger(typeof(AMQStateManager));

        /// <summary>
        /// The current state
        /// </summary>
        private AMQState _currentState;

        /// <summary>
        /// Maps from an AMQState instance to a Map from Class to StateTransitionHandler.
        /// The class must be a subclass of AMQFrame.
        /// </summary>
        private readonly IDictionary _state2HandlersMap = new Hashtable();

        private CopyOnWriteArraySet _stateListeners = new CopyOnWriteArraySet();
        
        public AMQStateManager()
        {
            _currentState = AMQState.CONNECTION_NOT_STARTED;
            RegisterListeners();
        }

        private void RegisterListeners()
        {
            IDictionary frame2handlerMap = new Hashtable();

            // we need to register a map for the null (i.e. all state) handlers otherwise you get
            // a stack overflow in the handler searching code when you present it with a frame for which
            // no handlers are registered
            //
            _state2HandlersMap[AMQState.ALL] = frame2handlerMap;

            frame2handlerMap = new Hashtable();
            frame2handlerMap[typeof(ConnectionStartBody)] = ConnectionStartMethodHandler.GetInstance();
            frame2handlerMap[typeof(ConnectionCloseBody)] = ConnectionCloseMethodHandler.GetInstance();
            _state2HandlersMap[AMQState.CONNECTION_NOT_STARTED] = frame2handlerMap;

            frame2handlerMap = new Hashtable();
            frame2handlerMap[typeof(ConnectionTuneBody)] = ConnectionTuneMethodHandler.GetInstance();
            frame2handlerMap[typeof(ConnectionSecureBody)] = ConnectionSecureMethodHandler.GetInstance();
            frame2handlerMap[typeof(ConnectionCloseBody)] = ConnectionCloseMethodHandler.GetInstance();
            _state2HandlersMap[AMQState.CONNECTION_NOT_TUNED] = frame2handlerMap;

            frame2handlerMap = new Hashtable();
            frame2handlerMap[typeof(ConnectionOpenOkBody)] = ConnectionOpenOkMethodHandler.GetInstance();
            frame2handlerMap[typeof(ConnectionCloseBody)] = ConnectionCloseMethodHandler.GetInstance();
            _state2HandlersMap[AMQState.CONNECTION_NOT_OPENED] = frame2handlerMap;

            //
            // ConnectionOpen handlers
            //
            frame2handlerMap = new Hashtable();
            frame2handlerMap[typeof(ChannelCloseBody)] = ChannelCloseMethodHandler.GetInstance();
            frame2handlerMap[typeof(ConnectionCloseBody)] = ConnectionCloseMethodHandler.GetInstance();
            frame2handlerMap[typeof(BasicDeliverBody)] = BasicDeliverMethodHandler.GetInstance();
            frame2handlerMap[typeof(BasicReturnBody)] = BasicReturnMethodHandler.GetInstance();
            _state2HandlersMap[AMQState.CONNECTION_OPEN] = frame2handlerMap;
        }

        public AMQState CurrentState
        {
            get
            {
                return _currentState;
            }
        }

        /// <summary>
        /// Changes the state.
        /// </summary>
        /// <param name="newState">The new state.</param>
        /// <exception cref="AMQException">if there is an error changing state</exception>
        public void ChangeState(AMQState newState)
        {
            _logger.Debug("State changing to " + newState + " from old state " + _currentState);
            AMQState oldState = _currentState;
            _currentState = newState;

            foreach (IStateListener l in _stateListeners)
            {
                l.StateChanged(oldState, newState);
            }
        }

        public void Error(AMQException e)
        {
            _logger.Debug("State manager receive error notification: " + e);
            foreach (IStateListener l in _stateListeners)
            {
                l.Error(e);
            }
        }

        public bool MethodReceived(AMQMethodEvent evt)
        {
            IStateAwareMethodListener handler = FindStateTransitionHandler(_currentState, evt.Method);
            if (handler != null)
            {
                handler.MethodReceived(this, evt);
                return true;
            }
            return false;
        }

        /// <summary>
        /// Finds the state transition handler.
        /// </summary>
        /// <param name="currentState">State of the current.</param>
        /// <param name="frame">The frame.</param>
        /// <returns></returns>
        /// <exception cref="IllegalStateTransitionException">if the state transition if not allowed</exception>
        private IStateAwareMethodListener FindStateTransitionHandler(AMQState currentState,
                                                                     AMQMethodBody frame)                
        {
            Type clazz = frame.GetType();
            if (_logger.IsDebugEnabled)
            {
                _logger.Debug("Looking for state transition handler for frame " + clazz);
            }
            IDictionary classToHandlerMap = (IDictionary) _state2HandlersMap[currentState];

            if (classToHandlerMap == null)
            {
                // if no specialised per state handler is registered look for a
                // handler registered for "all" states
                return FindStateTransitionHandler(AMQState.ALL, frame);
            }
            IStateAwareMethodListener handler = (IStateAwareMethodListener) classToHandlerMap[clazz];
            if (handler == null)
            {
                if (currentState == AMQState.ALL)
                {
                    _logger.Debug("No state transition handler defined for receiving frame " + frame);
                    return null;
                }
                else
                {
                    // if no specialised per state handler is registered look for a
                    // handler registered for "all" states
                    return FindStateTransitionHandler(AMQState.ALL, frame);
                }
            }
            else
            {
                return handler;
            }
        }

        public void AddStateListener(IStateListener listener)
        {
            _logger.Debug("Adding state listener");
            _stateListeners.Add(listener);
        }

        public void RemoveStateListener(IStateListener listener)
        {
            _stateListeners.Remove(listener);
        }

        public void AttainState(AMQState s)
        {
            if (_currentState != s)
            {
                _logger.Debug("Adding state wait to reach state " + s);
                StateWaiter sw = new StateWaiter(s);
                AddStateListener(sw);
                sw.WaituntilStateHasChanged();
                // at this point the state will have changed.
            }
        }        
    }
}

File: java/src/org/openamq/client/state/AMQStateManager.java
package org.openamq.client.state;

import edu.emory.mathcs.backport.java.util.concurrent.CopyOnWriteArraySet;
import org.apache.log4j.Logger;
import org.openamq.AMQException;
import org.openamq.framing.AMQMethodBody;
import org.openamq.client.protocol.*;
import org.openamq.client.handler.*;
import org.openamq.framing.*;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * The state manager is responsible for managing the state of the protocol session.
 * <p/>
 * For each AMQProtocolHandler there is a separate state manager.
 *
 */
public class AMQStateManager implements AMQMethodListener
{
    private static final Logger _logger = Logger.getLogger(AMQStateManager.class);

    /**
     * The current state
     */
    private AMQState _currentState;

    /**
     * Maps from an AMQState instance to a Map from Class to StateTransitionHandler.
     * The class must be a subclass of AMQFrame.
     */
    private final Map _state2HandlersMap = new HashMap();

    private final CopyOnWriteArraySet _stateListeners = new CopyOnWriteArraySet();

    public AMQStateManager()
    {
        this(AMQState.CONNECTION_NOT_STARTED, true);
    }

    protected AMQStateManager(AMQState state, boolean register)
    {
        _currentState = state;
        if(register)
        {
            registerListeners();
        }
    }

    protected void registerListeners()
    {
        Map frame2handlerMap = new HashMap();

        // we need to register a map for the null (i.e. all state) handlers otherwise you get
        // a stack overflow in the handler searching code when you present it with a frame for which
        // no handlers are registered
        //
        _state2HandlersMap.put(null, frame2handlerMap);

        frame2handlerMap = new HashMap();
        frame2handlerMap.put(ConnectionStartBody.class, ConnectionStartMethodHandler.getInstance());
        frame2handlerMap.put(ConnectionCloseBody.class, ConnectionCloseMethodHandler.getInstance());
        _state2HandlersMap.put(AMQState.CONNECTION_NOT_STARTED, frame2handlerMap);

        frame2handlerMap = new HashMap();
        frame2handlerMap.put(ConnectionTuneBody.class, ConnectionTuneMethodHandler.getInstance());
        frame2handlerMap.put(ConnectionSecureBody.class, ConnectionSecureMethodHandler.getInstance());
        frame2handlerMap.put(ConnectionCloseBody.class, ConnectionCloseMethodHandler.getInstance());
        _state2HandlersMap.put(AMQState.CONNECTION_NOT_TUNED, frame2handlerMap);

        frame2handlerMap = new HashMap();
        frame2handlerMap.put(ConnectionOpenOkBody.class, ConnectionOpenOkMethodHandler.getInstance());
        frame2handlerMap.put(ConnectionCloseBody.class, ConnectionCloseMethodHandler.getInstance());
        _state2HandlersMap.put(AMQState.CONNECTION_NOT_OPENED, frame2handlerMap);

        //
        // ConnectionOpen handlers
        //
        frame2handlerMap = new HashMap();
        frame2handlerMap.put(ChannelCloseBody.class, ChannelCloseMethodHandler.getInstance());
        frame2handlerMap.put(ConnectionCloseBody.class, ConnectionCloseMethodHandler.getInstance());
        frame2handlerMap.put(BasicDeliverBody.class, BasicDeliverMethodHandler.getInstance());
        frame2handlerMap.put(BasicReturnBody.class, BasicReturnMethodHandler.getInstance());
        frame2handlerMap.put(ChannelFlowOkBody.class, ChannelFlowOkMethodHandler.getInstance());
        _state2HandlersMap.put(AMQState.CONNECTION_OPEN, frame2handlerMap);
    }

    public AMQState getCurrentState()
    {
        return _currentState;
    }

    public void changeState(AMQState newState) throws AMQException
    {
        _logger.debug("State changing to " + newState + " from old state " + _currentState);
        final AMQState oldState = _currentState;
        _currentState = newState;

        synchronized (_stateListeners)
        {
            final Iterator it = _stateListeners.iterator();
            while (it.hasNext())
            {
                final StateListener l = (StateListener) it.next();
                l.stateChanged(oldState, newState);
            }
        }
    }

    public void error(Exception e)
    {
        _logger.debug("State manager receive error notification: " + e);
        synchronized (_stateListeners)
        {
            final Iterator it = _stateListeners.iterator();
            while (it.hasNext())
            {
                final StateListener l = (StateListener) it.next();
                l.error(e);
            }
        }
    }

    public boolean methodReceived(AMQMethodEvent evt) throws AMQException
    {
        StateAwareMethodListener handler = findStateTransitionHandler(_currentState, evt.getMethod());
        if (handler != null)
        {
            handler.methodReceived(this, evt);
            return true;
        }
        return false;
    }

    protected StateAwareMethodListener findStateTransitionHandler(AMQState currentState,
                                                               AMQMethodBody frame)
            throws IllegalStateTransitionException
    {
        final Class clazz = frame.getClass();
        if (_logger.isDebugEnabled())
        {
            _logger.debug("Looking for state transition handler for frame " + clazz);
        }
        final Map classToHandlerMap = (Map) _state2HandlersMap.get(currentState);

        if (classToHandlerMap == null)
        {
            // if no specialised per state handler is registered look for a
            // handler registered for "all" states
            return findStateTransitionHandler(null, frame);
        }
        final StateAwareMethodListener handler = (StateAwareMethodListener) classToHandlerMap.get(clazz);
        if (handler == null)
        {
            if (currentState == null)
            {
                _logger.debug("No state transition handler defined for receiving frame " + frame);
                return null;
            }
            else
            {
                // if no specialised per state handler is registered look for a
                // handler registered for "all" states
                return findStateTransitionHandler(null, frame);
            }
        }
        else
        {
            return handler;
        }
    }

    public void addStateListener(StateListener listener)
    {
        _logger.debug("Adding state listener");
        _stateListeners.add(listener);
    }

    public void removeStateListener(StateListener listener)
    {
        _stateListeners.remove(listener);
    }

    public void attainState(AMQState s) throws AMQException
    {
        boolean needToWait = false;
        StateWaiter sw = null;
        synchronized (_stateListeners)
        {
            if (_currentState != s)
            {
                _logger.debug("Adding state wait to reach state " + s);
                sw = new StateWaiter(s);
                addStateListener(sw);
                // we use a boolean since we must release the lock before starting to wait
                needToWait = true;
            }
        }
        if (needToWait)
        {
            sw.waituntilStateHasChanged();
        }
        // at this point the state will have changed.
    }
}
